File: backends/node/src/schema.ts
/**
 * Wire schema for `agent_gate/v1`.
 *
 * Mirrors `GateRequest.toJson()` / `GateDecision.fromJson()` in the Flutter
 * package. Keep the two in sync — the Flutter side is the source of truth.
 * We validate strictly enough to reject garbage but stay permissive on
 * `context.app`, which is free-form by design.
 */
import { z } from "zod";

export const CandidateSchema = z.object({
  id: z.string().min(1),
  label: z.string(),
  description: z.string(),
  route: z.string().optional(),
  tags: z.array(z.string()).optional(),
  priority: z.number().int().optional(),
  metadata: z.record(z.unknown()).optional(),
});

export const PageSummarySchema = z.object({
  page: z.string(),
  entered_at: z.string().optional(),
  dwell_ms: z.number().optional(),
  taps: z.number().optional(),
  attempts: z.number().optional(),
  failures: z.number().optional(),
  successes: z.number().optional(),
  validation_errors: z.number().optional(),
  backs: z.number().optional(),
  field_edits: z.number().optional(),
  max_scroll: z.number().optional(),
  hesitation_ms: z.number().optional(),
  attempts_by_name: z.record(z.number()).optional(),
  taps_by_target: z.record(z.number()).optional(),
  events: z.number().optional(),
});

export const ContextSchema = z.object({
  consent: z.boolean().default(false),
  current_page: PageSummarySchema.optional(),
  history: z.array(PageSummarySchema).optional(),
  events: z.array(z.unknown()).optional(),
  app: z.record(z.unknown()).default({}),
  baseline: z.record(z.unknown()).optional(),
  device: z.record(z.unknown()).optional(),
});

export const GateRequestSchema = z.object({
  schema: z.literal("agent_gate/v1"),
  request_id: z.string().min(1),
  timestamp: z.string(),
  gate_id: z.string().min(1),
  from_page: z.string(),
  profile: z.enum(["risk", "recommendation", "general"]),
  instructions: z.string().optional(),
  candidates: z.array(CandidateSchema).min(1).max(1000),
  context: ContextSchema,
});

export type GateRequest = z.infer<typeof GateRequestSchema>;
export type Candidate = z.infer<typeof CandidateSchema>;

/** What we send back. Matches `GateDecision.fromJson`. */
export interface GateDecision {
  candidate_id: string;
  confidence: number; // 0..1
  reason: string;
  model?: string;
}

File: backends/python/app/schema.py
"""
Wire schema for ``agent_gate/v1``.

Mirrors ``GateRequest.toJson()`` / ``GateDecision.fromJson()`` in the Flutter
package (the source of truth). Strict where it matters (schema tag, candidate
ids, profile), permissive on ``context.app`` which is free-form by design.
"""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field, conlist


class Candidate(BaseModel):
    id: str = Field(min_length=1)
    label: str
    description: str
    route: str | None = None
    tags: list[str] = []
    priority: int = 0
    metadata: dict[str, Any] = {}


class PageSummary(BaseModel):
    page: str
    dwell_ms: int = 0
    taps: int = 0
    attempts: int = 0
    failures: int = 0
    successes: int = 0
    validation_errors: int = 0
    backs: int = 0
    field_edits: int = 0
    max_scroll: float = 0.0
    hesitation_ms: int | None = None
    attempts_by_name: dict[str, int] = {}
    taps_by_target: dict[str, int] = {}
    events: int = 0
    entered_at: str | None = None


class Context(BaseModel):
    consent: bool = False
    current_page: PageSummary | None = None
    history: list[PageSummary] = []
    events: list[Any] = []
    app: dict[str, Any] = {}
    baseline: dict[str, Any] | None = None
    device: dict[str, Any] = {}


class GateRequest(BaseModel):
    schema_: Literal["agent_gate/v1"] = Field(alias="schema")
    request_id: str = Field(min_length=1)
    timestamp: str
    gate_id: str = Field(min_length=1)
    from_page: str
    profile: Literal["risk", "recommendation", "general"]
    instructions: str | None = None
    candidates: conlist(Candidate, min_length=1, max_length=1000)  # type: ignore[valid-type]
    context: Context = Context()

    model_config = {"populate_by_name": True}

    def candidate_ids(self) -> list[str]:
        return [c.id for c in self.candidates]

    def has(self, cid: str) -> bool:
        return any(c.id == cid for c in self.candidates)

    def by_tag(self, tag: str) -> Candidate | None:
        return next((c for c in self.candidates if tag in c.tags), None)

    def default_candidate(self) -> Candidate:
        """First candidate tagged ``default`` (highest priority), else the first."""
        tagged = sorted((c for c in self.candidates if "default" in c.tags), key=lambda c: -c.priority)
        return tagged[0] if tagged else self.candidates[0]


class GateDecision(BaseModel):
    candidate_id: str
    confidence: float = Field(ge=0.0, le=1.0)
    reason: str
    model: str | None = None
